From 25be92f25a3ce6a2ecab30f35b1623601a926d40 Mon Sep 17 00:00:00 2001
From: Yushan Wang <wangyushan12@huawei.com>
Date: Thu, 17 Apr 2025 21:49:29 +0800
Subject: [PATCH 487/960] downstream: soc cache: Add framework driver for
 HiSilicon SoC cache

HiSilicon SoC cache is comprised of multiple hardware devices, a driver
in this patch is used to provide common utilities for other drivers to
avoid redundancy.

The driver will create a file of `/dev/hisi_soc_cache_mgmt`, mmap
operations to it will be taken as applying cache lock for a memory
region, ioctl operations will be taken as cache maintenance.

The address is aligned up to return to user and the arena's starting
address is stored for future release.  Unused pages produced during
above process are released to allow minimum arena.

Signed-off-by: Yicong Yang <yangyicong@hisilicon.com>
Signed-off-by: Yushan Wang <wangyushan12@huawei.com>
Signed-off-by: Jie Wang <wangjie125@huawei.com>
Signed-off-by: Slim6882 <15605196882@163.com>
Signed-off-by: chenyi <chenyi211@h-partners.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/soc/hisilicon/Kconfig                 |  11 +
 drivers/soc/hisilicon/Makefile                |   1 +
 .../soc/hisilicon/hisi_soc_cache_framework.c  | 378 ++++++++++++++++++
 .../soc/hisilicon/hisi_soc_cache_framework.h  |  77 ++++
 .../uapi/misc/hisi_soc_cache/hisi_soc_cache.h |  35 ++
 5 files changed, 502 insertions(+)
 create mode 100644 drivers/soc/hisilicon/hisi_soc_cache_framework.c
 create mode 100644 drivers/soc/hisilicon/hisi_soc_cache_framework.h
 create mode 100644 include/uapi/misc/hisi_soc_cache/hisi_soc_cache.h

diff --git a/drivers/soc/hisilicon/Kconfig b/drivers/soc/hisilicon/Kconfig
index 0ab688af308f..a85dbe8ac60c 100644
--- a/drivers/soc/hisilicon/Kconfig
+++ b/drivers/soc/hisilicon/Kconfig
@@ -18,4 +18,15 @@ config KUNPENG_HCCS
 	  Say M here if you want to include support for querying the
 	  health status and port information of HCCS on Kunpeng SoC.
 
+config HISI_SOC_CACHE
+	tristate "HiSilicon Cache driver for Kunpeng SoC"
+	depends on ARCH_HISI
+	help
+	  This driver provides the basic utilities for drivers of
+	  different part of Kunpeng SoC cache, including L3 cache and
+	  Hydra Home Agent etc.
+
+	  If either HiSilicon L3 cache driver or HiSilicon Hydra Home
+	  Agent driver is needed, say yes.
+
 endmenu
diff --git a/drivers/soc/hisilicon/Makefile b/drivers/soc/hisilicon/Makefile
index 226e747e70d6..dbc673b71d4d 100644
--- a/drivers/soc/hisilicon/Makefile
+++ b/drivers/soc/hisilicon/Makefile
@@ -1,2 +1,3 @@
 # SPDX-License-Identifier: GPL-2.0-only
 obj-$(CONFIG_KUNPENG_HCCS)	+= kunpeng_hccs.o
+obj-$(CONFIG_HISI_SOC_CACHE)	+= hisi_soc_cache_framework.o
diff --git a/drivers/soc/hisilicon/hisi_soc_cache_framework.c b/drivers/soc/hisilicon/hisi_soc_cache_framework.c
new file mode 100644
index 000000000000..1782e6a44668
--- /dev/null
+++ b/drivers/soc/hisilicon/hisi_soc_cache_framework.c
@@ -0,0 +1,378 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Framework for HiSilicon SoC cache, manages HiSilicon SoC cache drivers.
+ *
+ * Copyright (c) 2024 HiSilicon Technologies Co., Ltd.
+ * Author: Jie Wang <wangjie125@huawei.com>
+ * Author: Yicong Yang <yangyicong@hisilicon.com>
+ * Author: Yushan Wang <wangyushan12@huawei.com>
+ */
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+
+#include <linux/cleanup.h>
+#include <linux/debugfs.h>
+#include <linux/fs.h>
+#include <linux/init.h>
+#include <linux/ioctl.h>
+#include <linux/memory.h>
+#include <linux/miscdevice.h>
+#include <linux/module.h>
+#include <linux/pagewalk.h>
+#include <linux/spinlock.h>
+#include <linux/uaccess.h>
+
+#include <asm/page.h>
+
+#include "hisi_soc_cache_framework.h"
+
+struct hisi_soc_comp_inst {
+	struct list_head node;
+	struct hisi_soc_comp *comp;
+};
+
+struct hisi_soc_comp_list {
+	struct list_head node;
+	/* protects list of HiSilicon SoC cache components */
+	spinlock_t lock;
+	u32 inst_num;
+};
+
+static struct hisi_soc_comp_list soc_cache_devs[SOC_COMP_TYPE_MAX];
+
+int hisi_soc_cache_maintain(phys_addr_t addr, size_t size,
+			    enum hisi_soc_cache_maint_type mnt_type)
+{
+	struct hisi_soc_comp_inst *inst;
+	struct list_head *head;
+	int ret = -EOPNOTSUPP;
+
+	if (mnt_type >= HISI_CACHE_MAINT_MAX)
+		return -EINVAL;
+
+	guard(spinlock)(&soc_cache_devs[HISI_SOC_HHA].lock);
+
+	head = &soc_cache_devs[HISI_SOC_HHA].node;
+	list_for_each_entry(inst, head, node) {
+		ret = inst->comp->ops->do_maintain(inst->comp, addr, size,
+						   mnt_type);
+		if (ret)
+			return ret;
+	}
+
+	list_for_each_entry(inst, head, node) {
+		ret = inst->comp->ops->poll_maintain_done(inst->comp, addr,
+							  size, mnt_type);
+		if (ret)
+			return ret;
+	}
+
+	return ret;
+}
+EXPORT_SYMBOL_GPL(hisi_soc_cache_maintain);
+
+static int hisi_soc_cache_maint_pte_entry(pte_t *pte, unsigned long addr,
+				unsigned long next, struct mm_walk *walk)
+{
+#ifdef HISI_SOC_CACHE_LLT
+	unsigned int mnt_type = *((unsigned int *)walk->priv);
+#else
+	unsigned int mnt_type = *((unsigned int *)walk->private);
+#endif
+	size_t size = next - addr;
+	phys_addr_t paddr;
+
+	if (!pte_present(ptep_get(pte)))
+		return -EINVAL;
+
+	paddr = PFN_PHYS(pte_pfn(*pte)) + offset_in_page(addr);
+
+	return hisi_soc_cache_maintain(paddr, size, mnt_type);
+}
+
+static const struct mm_walk_ops hisi_soc_cache_maint_walk = {
+	.pte_entry = hisi_soc_cache_maint_pte_entry,
+	.walk_lock = PGWALK_RDLOCK,
+};
+
+static int hisi_soc_cache_inst_check(const struct hisi_soc_comp *comp,
+				     enum hisi_soc_comp_type comp_type)
+{
+	/* Different types of component could have different ops. */
+	switch (comp_type) {
+	case HISI_SOC_HHA:
+		if (!comp->ops->do_maintain || !comp->ops->poll_maintain_done)
+			return -EINVAL;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int hisi_soc_cache_inst_add(struct hisi_soc_comp *comp,
+				   enum hisi_soc_comp_type comp_type)
+{
+	struct hisi_soc_comp_inst *comp_inst;
+	int ret;
+
+	ret = hisi_soc_cache_inst_check(comp, comp_type);
+	if (ret)
+		return ret;
+
+	comp_inst = kzalloc(sizeof(*comp_inst), GFP_KERNEL);
+	if (!comp_inst)
+		return -ENOMEM;
+
+	comp_inst->comp = comp;
+
+	scoped_guard(spinlock, &soc_cache_devs[comp_type].lock) {
+		list_add_tail(&comp_inst->node,
+			      &soc_cache_devs[comp_type].node);
+		soc_cache_devs[comp_type].inst_num++;
+	}
+
+	return 0;
+}
+
+/*
+ * When @comp is NULL, it means to delete all instances of @comp_type.
+ */
+static void hisi_soc_cache_inst_del(struct hisi_soc_comp *comp,
+				    enum hisi_soc_comp_type comp_type)
+{
+	struct hisi_soc_comp_inst *inst, *tmp;
+
+	guard(spinlock)(&soc_cache_devs[comp_type].lock);
+	list_for_each_entry_safe(inst, tmp, &soc_cache_devs[comp_type].node,
+				 node) {
+		if (comp && comp != inst->comp)
+			continue;
+
+		if (soc_cache_devs[comp_type].inst_num > 0)
+			soc_cache_devs[comp_type].inst_num--;
+
+		list_del(&inst->node);
+		kfree(inst);
+
+		/* Stop the loop if we have already deleted @comp. */
+		if (comp)
+			break;
+	}
+}
+
+int hisi_soc_comp_inst_add(struct hisi_soc_comp *comp)
+{
+	int ret, i = 0;
+
+	if (!comp || !comp->ops || comp->comp_type == 0)
+		return -EINVAL;
+
+	for_each_set_bit_from(i, &comp->comp_type, SOC_COMP_TYPE_MAX) {
+		ret = hisi_soc_cache_inst_add(comp, i);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(hisi_soc_comp_inst_add);
+
+int hisi_soc_comp_inst_del(struct hisi_soc_comp *comp)
+{
+	int i;
+
+	if (!comp)
+		return -EINVAL;
+
+	for_each_set_bit(i, &comp->comp_type, SOC_COMP_TYPE_MAX)
+		hisi_soc_cache_inst_del(comp, i);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(hisi_soc_comp_inst_del);
+
+static int __hisi_soc_cache_maintain(unsigned long __user vaddr, size_t size,
+				     enum hisi_soc_cache_maint_type mnt_type)
+{
+	unsigned long start = untagged_addr(vaddr);
+	struct vm_area_struct *vma;
+	int ret = 0;
+
+	mmap_read_lock_killable(current->mm);
+
+	vma = vma_lookup(current->mm, vaddr);
+	if (!vma || vaddr + size > vma->vm_end || !size) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* User should have the write permission of target memory */
+	if (!(vma->vm_flags & VM_WRITE)) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	ret = walk_page_range(current->mm, start, start + size,
+			&hisi_soc_cache_maint_walk, &mnt_type);
+
+out:
+	mmap_read_unlock(current->mm);
+	return ret;
+}
+
+static long hisi_soc_cache_mgmt_ioctl(struct file *file, u32 cmd, unsigned long arg)
+{
+	struct hisi_soc_cache_ioctl_param *param =
+		kzalloc(sizeof(struct hisi_soc_cache_ioctl_param), GFP_KERNEL);
+	long ret;
+
+	if (!param) {
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	if (copy_from_user(param, (void __user *)arg, sizeof(*param))) {
+		ret = -EFAULT;
+		goto out;
+	}
+
+	switch (cmd) {
+	case HISI_CACHE_MAINTAIN:
+		ret = __hisi_soc_cache_maintain(param->addr, param->size,
+						param->op_type);
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	}
+out:
+	kfree(param);
+	return ret;
+}
+
+static const struct file_operations soc_cache_dev_fops = {
+	.owner = THIS_MODULE,
+	.unlocked_ioctl = hisi_soc_cache_mgmt_ioctl,
+};
+
+static struct miscdevice soc_cache_miscdev = {
+	.minor = MISC_DYNAMIC_MINOR,
+	.name = "hisi_soc_cache_mgmt",
+	.fops = &soc_cache_dev_fops,
+	.mode = 0600,
+};
+
+static void hisi_soc_cache_inst_uninit(void)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(soc_cache_devs); ++i)
+		hisi_soc_cache_inst_del(NULL, i);
+}
+
+static void hisi_soc_cache_framework_data_init(void)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(soc_cache_devs); ++i) {
+		spin_lock_init(&soc_cache_devs[i].lock);
+		INIT_LIST_HEAD(&soc_cache_devs[i].node);
+	}
+}
+
+static const char *const hisi_soc_cache_item_str[SOC_COMP_TYPE_MAX] = {
+	"hha"
+};
+
+/*
+ * Print cache instance number debug information for debug FS.
+ */
+static ssize_t hisi_soc_cache_dbg_get_inst_num(struct file *file,
+					       char __user *buff,
+					       size_t cnt,
+					       loff_t *ppos)
+{
+#define HISI_SOC_CACHE_DBGFS_REG_LEN 100
+	char *read_buff;
+	int len, i, pos = 0;
+	int ret = 0;
+
+	if (!access_ok(buff, cnt))
+		return -EFAULT;
+	if (*ppos < 0)
+		return -EINVAL;
+	if (cnt == 0)
+		return 0;
+
+	read_buff = kzalloc(HISI_SOC_CACHE_DBGFS_REG_LEN, GFP_KERNEL);
+	if (!read_buff)
+		return -ENOMEM;
+
+	len = HISI_SOC_CACHE_DBGFS_REG_LEN;
+
+	for (i = 0; i < ARRAY_SIZE(soc_cache_devs); i++) {
+		guard(spinlock)(&soc_cache_devs[i].lock);
+		pos += scnprintf(read_buff + pos, len - pos,
+				 "%s inst num: %u\n",
+				 hisi_soc_cache_item_str[i],
+				 soc_cache_devs[i].inst_num);
+	}
+
+	ret = simple_read_from_buffer(buff, cnt, ppos, read_buff,
+				       strlen(read_buff));
+	kfree(read_buff);
+	return ret;
+}
+
+static struct dentry *hisi_cache_dbgfs_root;
+static const struct file_operations hisi_cache_dbgfs_ops = {
+	.owner = THIS_MODULE,
+	.open = simple_open,
+	.read = hisi_soc_cache_dbg_get_inst_num,
+};
+
+static void hisi_soc_cache_dbgfs_init(void)
+{
+	hisi_cache_dbgfs_root = debugfs_create_dir("hisi_soc_cache_frm", NULL);
+	debugfs_create_file("instance", 0400, hisi_cache_dbgfs_root, NULL,
+			    &hisi_cache_dbgfs_ops);
+}
+
+static void hisi_soc_cache_dbgfs_uninit(void)
+{
+	debugfs_remove_recursive(hisi_cache_dbgfs_root);
+	hisi_cache_dbgfs_root = NULL;
+}
+
+static int __init hisi_soc_cache_framework_init(void)
+{
+	int ret;
+
+	hisi_soc_cache_framework_data_init();
+
+	ret = misc_register(&soc_cache_miscdev);
+	if (ret) {
+		hisi_soc_cache_inst_uninit();
+		return ret;
+	}
+
+	hisi_soc_cache_dbgfs_init();
+
+	return 0;
+}
+module_init(hisi_soc_cache_framework_init);
+
+static void __exit hisi_soc_cache_framework_exit(void)
+{
+	hisi_soc_cache_dbgfs_uninit();
+	misc_deregister(&soc_cache_miscdev);
+	hisi_soc_cache_inst_uninit();
+}
+module_exit(hisi_soc_cache_framework_exit);
+
+MODULE_DESCRIPTION("HiSilicon SoC Cache Framework Driver");
+MODULE_AUTHOR("Jie Wang <wangjie125@huawei.com>");
+MODULE_AUTHOR("Yushan Wang <wangyushan12@huawei.com>");
+MODULE_LICENSE("GPL");
diff --git a/drivers/soc/hisilicon/hisi_soc_cache_framework.h b/drivers/soc/hisilicon/hisi_soc_cache_framework.h
new file mode 100644
index 000000000000..9b3de4e50161
--- /dev/null
+++ b/drivers/soc/hisilicon/hisi_soc_cache_framework.h
@@ -0,0 +1,77 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Header file of framework for HiSilicon SoC cache.
+ *
+ * Copyright (c) 2024 HiSilicon Technologies Co., Ltd.
+ * Author: Jie Wang <wangjie125@huawei.com>
+ * Author: Yicong Yang <yangyicong@hisilicon.com>
+ * Author: Yushan Wang <wangyushan12@huawei.com>
+ */
+
+#ifndef HISI_CACHE_FRAMEWORK_H
+#define HISI_CACHE_FRAMEWORK_H
+
+#include <linux/bits.h>
+#include <linux/types.h>
+
+#include <uapi/misc/hisi_soc_cache/hisi_soc_cache.h>
+
+enum hisi_soc_comp_type {
+	HISI_SOC_HHA,
+	SOC_COMP_TYPE_MAX
+};
+
+struct hisi_soc_comp;
+
+/**
+ * struct hisi_soc_comp_ops - Callbacks for SoC cache drivers to handle
+ *			      operation requests.
+ * @maintain_enable: perform certain cache maintain operation on HHA.
+ * @poll_maintain_done: check if the HHA maintain operation has succeeded.
+ *
+ * Operations are decoupled into two phases so that framework does not have
+ * to wait for one operation to finish before calling the next when multiple
+ * hardwares onboard.
+ *
+ * Implementers must implement the functions in pairs.  Implementation should
+ * return -EBUSY when:
+ * - insufficient resources are available to perform the operation.
+ * - previously raised operation is not finished.
+ * - new operations (do_lock(), do_unlock() etc.) to the same address
+ *   before corresponding done functions being called.
+ */
+struct hisi_soc_comp_ops {
+	int (*do_maintain)(struct hisi_soc_comp *comp,
+			      phys_addr_t addr, size_t size,
+			      enum hisi_soc_cache_maint_type mnt_type);
+	int (*poll_maintain_done)(struct hisi_soc_comp *comp,
+			      phys_addr_t addr, size_t size,
+			      enum hisi_soc_cache_maint_type mnt_type);
+};
+
+/**
+ * struct hisi_soc_comp - Struct of HiSilicon SoC cache components.
+ * @ops: possible operations a component may perform.
+ * @affinity_mask: cpus that associate with this component.
+ * @comp_type: bitmap declaring the type of the component.
+ *
+ * A component may have multiple types (e.g. a piece of multi-function device).
+ * If so, set the bit of @comp_type according to its supporting type in struct
+ * hisi_soc_comp_type.
+ */
+struct hisi_soc_comp {
+	struct hisi_soc_comp_ops *ops;
+	cpumask_t affinity_mask;
+	/*
+	 * Setting bit x to 1 means this instance supports feature of x-th
+	 * entry in enum hisi_soc_comp_type.
+	 */
+	unsigned long comp_type;
+};
+
+int hisi_soc_comp_inst_add(struct hisi_soc_comp *comp);
+int hisi_soc_comp_inst_del(struct hisi_soc_comp *comp);
+int hisi_soc_cache_maintain(phys_addr_t addr, size_t size,
+			    enum hisi_soc_cache_maint_type mnt_type);
+
+#endif
diff --git a/include/uapi/misc/hisi_soc_cache/hisi_soc_cache.h b/include/uapi/misc/hisi_soc_cache/hisi_soc_cache.h
new file mode 100644
index 000000000000..5441f6f75b81
--- /dev/null
+++ b/include/uapi/misc/hisi_soc_cache/hisi_soc_cache.h
@@ -0,0 +1,35 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later WITH Linux-syscall-note */
+/* Copyright (c) 2024-2024 HiSilicon Limited. */
+#ifndef _UAPI_HISI_SOC_CACHE_H
+#define _UAPI_HISI_SOC_CACHE_H
+
+#include <linux/types.h>
+
+/* HISI_CACHE_MAINTAIN: cache maintain operation for HiSilicon SoC */
+#define HISI_CACHE_MAINTAIN	_IOW('C', 1, unsigned long)
+
+/*
+ * Further information of these operations can be found at:
+ * https://developer.arm.com/documentation/ihi0050/latest/
+ */
+enum hisi_soc_cache_maint_type {
+	HISI_CACHE_MAINT_CLEANSHARED,
+	HISI_CACHE_MAINT_CLEANINVALID,
+	HISI_CACHE_MAINT_MAKEINVALID,
+
+	HISI_CACHE_MAINT_MAX
+};
+
+/**
+ * struct hisi_soc_cache_ioctl_param - User data for hisi cache operates.
+ * @op_type: cache maintain type
+ * @addr: cache maintain address
+ * @size: cache maintain size
+ */
+struct hisi_soc_cache_ioctl_param {
+	enum hisi_soc_cache_maint_type op_type;
+	unsigned long addr;
+	unsigned long size;
+};
+
+#endif
-- 
2.43.0

